File: server/src/main/java/model/item/Equipment.java
package model.item;

import javafx.beans.property.IntegerProperty;
import javafx.beans.property.ReadOnlyIntegerProperty;
import javafx.beans.property.SimpleIntegerProperty;
import manager.Database;
import manager.TypTranslation;
import manager.Utility;
import model.ICurrency;
import model.upgrade.IUpgrade;
import model.upgrade.UpgradeFactory;

import java.util.*;
import java.util.function.Consumer;
import java.util.stream.Collectors;

public abstract class Equipment extends Item implements IEquipment {

    protected static Random rand = new Random();

    protected String material;
    protected int upgradeSlots;
    protected ArrayList<IUpgrade> upgrades;
    protected IntegerProperty wearStage;
    protected IntegerProperty wearTick;
    protected Collection<Consumer<Equipment>> onBreakListeners;

    public Equipment() {
        super();
        this.material = "";
        this.upgradeSlots = 0;
        this.upgrades = new ArrayList<>();
        this.wearTick = new SimpleIntegerProperty(0);
        this.wearStage = new SimpleIntegerProperty(0);

        int neededSteps = Utility.getConfig().getInt("character.wear.stepsNeeded");

        if (neededSteps > 0) {
            this.wearStage.bind(wearTick.divide(neededSteps));
        }
        this.onBreakListeners = new ArrayList<>();
    }

    protected abstract boolean shouldBreak();

    /**
     * Applies wear to this equipment
     *
     * @param wear the amount of wear
     */
    public void applyWear(int wear) {
        this.setWearTick(getWearTick() + wear);
        if (shouldBreak()) {
            for (Consumer<Equipment> listener : this.onBreakListeners) {
                listener.accept(this);
            }
        }
    }

    public Equipment copy() {
        Equipment equipment = (Equipment) super.copy();
        equipment.setMaterial(this.getMaterial());
        equipment.setUpgradeSlots(this.getUpgradeSlots());

        return equipment;
    }

    public Equipment getWithUpgrade() {

        Equipment equipment = this.copy();
        Collection<String> types = TypTranslation.getAllTypes(this.getSubtype());
        List<UpgradeFactory> list = Database.upgradeList.stream().
                filter(x -> types.contains(x.getTarget())).collect(Collectors.toList());

        if (list.size() == 0) {
            return equipment;
        }

        while (rand.nextDouble() <= 0.2) {
            UpgradeFactory factory = list.get(rand.nextInt(list.size()));
            if (factory.getSlots() <= equipment.getUpgradeSlots()) {
                equipment.getUpgrades().add(factory.getUpgrade());
                equipment.setUpgradeSlots(getUpgradeSlots() - factory.getSlots());
                list.remove(factory);
            }
        }

        return equipment;
    }

    public void addOnBreakListener(Consumer<Equipment> listener) {
        this.onBreakListeners.add(listener);
    }

    public void removeOnBreakListener(Consumer<Equipment> listener) {
        this.onBreakListeners.remove(listener);
    }

    public String getMaterial() {
        return material;
    }

    public void setMaterial(String material) {
        this.material = material;
    }

    public int getUpgradeSlots() {
        return upgradeSlots;
    }

    public void setUpgradeSlots(int upgradeSlots) {
        this.upgradeSlots = upgradeSlots;
    }

    public Collection<IUpgrade> getUpgrades() {
        return upgrades;
    }

    public void setUpgrades(ArrayList<IUpgrade> upgrades) {
        this.upgrades = upgrades;
        this.upgrades.sort(Comparator.comparing(IUpgrade::getFullName));
    }

    @Override
    public ICurrency getCurrency() {
        ICurrency currency = super.getCurrency();

        for (IUpgrade upgrade : upgrades) {
            currency = currency.add(upgrade.getCost());
        }

        return currency;
    }

    public int getWearStage() {
        return wearStage.get();
    }

    public ReadOnlyIntegerProperty wearStageProperty() {
        return wearStage;
    }

    public int getWearTick() {
        return wearTick.get();
    }

    public IntegerProperty wearTickProperty() {
        return wearTick;
    }

    public void setWearTick(int wearTick) {
        this.wearTick.set(wearTick);
    }

    @Override
    public boolean equals(Object o) {

        if (!(o instanceof Equipment) || !super.equals(o)) {
            return false;
        }

        Equipment other = (Equipment) o;

        this.upgrades.sort(Comparator.comparing(IUpgrade::getFullName));
        other.upgrades.sort(Comparator.comparing(IUpgrade::getFullName));

        return this.getMaterial().equals(other.getMaterial()) && this.getUpgradeSlots() == other.getUpgradeSlots() &&
                this.upgradesAsString().equals(other.upgradesAsString());
    }
}

File: app/src/main/java/de/pnp/manager/app/model/Equipment.kt
package de.pnp.manager.app.model

import de.pnp.manager.model.ICurrency
import de.pnp.manager.model.item.IEquipment
import de.pnp.manager.model.upgrade.IUpgrade

open class Equipment : IEquipment, Item() {
    private var wearTick : Int = 0
    private var wearStage : Int = 0
    private var material : String = ""
    private var upgrades : ArrayList<IUpgrade> = ArrayList()
    private var upgradeSlots : Int = 0

    override fun getCurrency(): ICurrency? {
        var currency = super.getCurrency()

        for (upgrade in upgrades) {
            currency = currency!!.add(upgrade.cost)
        }

        return currency
    }

    override fun copy(): IEquipment {
        TODO("Not yet implemented")
    }

    override fun applyWear(wear: Int) {
        setWearTick(getWearTick() + wear)
/*        if (shouldBreak()) {
            for (listener in this.onBreakListeners) {
                listener.accept(this)
            }
        }*/
    }

    override fun getWithUpgrade(): IEquipment {
        TODO("Not yet implemented")
    }

    override fun getMaterial(): String {
        return material
    }

    override fun setMaterial(material: String) {
        this.material = material
    }

    override fun getUpgradeSlots(): Int {
        return upgradeSlots
    }

    override fun setUpgradeSlots(upgradeSlots: Int) {
        this.upgradeSlots = upgradeSlots
    }

    override fun getUpgrades(): MutableCollection<IUpgrade> {
        return upgrades
    }

    override fun setUpgrades(upgrades: ArrayList<IUpgrade>) {
        this.upgrades = upgrades
    }

    override fun getWearStage(): Int {
        return wearStage
    }

    fun setWearStage(value: Int) {
        this.wearStage = value
    }

    override fun getWearTick(): Int {
        return wearTick
    }

    override fun setWearTick(wearTick: Int) {
        this.wearTick = wearTick
    }
}